### PR TITLE
Backport Go modules to release-0.1

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by ncdc** [comment]
Please change v1alpha1 to release-0.1 in the title. Thank you!

**Comment by vincepri** [comment]
done!

File: hack/ensure-go.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```
Comment (chuckha) [inline hack/ensure-go.sh]: chmod +x this bad boy

**Comment by detiber** [comment]
I'm really hesitant about backporting go modules to `release-0.1` since it will impact the ability of existing providers to consume updates from the `release-0.1` branch without also adopting go modules at the same time. I think it is better to make that transition with the larger v1alpha2 changes rather than forcing it as a patch release update.

That said, it shouldn't prevent the use of go modules for downstream providers, since it should read the dep files from cluster-api.

**Comment by vincepri** [comment]
I mentioned it yesterday in Slack but it's frustrating to work between repositories that have Go modules enabled and ones that don't. In addition, Go 1.13 (due next month) turns on modules by default. Providers will need to update regardless.

File: Dockerfile [hunk 2]
```diff
@@ -13,16 +13,20 @@
...
 # Build the manager binary
-FROM golang:1.12.5 as builder
+FROM golang:1.12.6 as builder
 
 # Copy in the go src
 WORKDIR ${GOPATH}/src/sigs.k8s.io/cluster-api
 COPY pkg/    pkg/
 COPY cmd/    cmd/
 COPY vendor/ vendor/
+COPY go.mod go.mod
```
[the remark this answers is not in the document]
Comment (vincepri) [reply Dockerfile]: Let's open an issue around this? I mostly copied the one that's currently in master

File: hack/ensure-go.sh (hunk 1 above)
Comment (vincepri) [reply hack/ensure-go.sh]: Done!

**Comment by detiber** [comment]
@vincepri We should match Go minor versions with the vendored version of Kubernetes we are using, which for v1alpha1 will remain `1.12.x`, so I don't think that `1.13` factors into the decision.

In this case, I think the pain for developers `cluster-api` against the `release-0.1` branch is a necessary requirement, since the alternative would cause breakage for existing provider implementations (not all of which are under sig-cluster-lifecycle or the greater Kubernetes umbrella).

**Comment by detiber** [comment]
After discussing further with @vincepri I'm good with backport go modules to `release-0.1` as long as we ensure that we continue to publish and update a functional Gopkg.toml for downstream providers.

**Comment by vincepri** [comment]
Added `Gopkg.lock` and `Gopkg.toml` back for backward-compatibility, I'll follow-up with another PR against master around backporting changes to `release-0.1`. How does that sound?

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	k8s.io/gengo v0.0.0-20190116091435-f8a0810f38af // indirect
+	k8s.io/klog v0.3.2
+	k8s.io/kube-openapi v0.0.0-20190228160746-b3a7cee44a30 // indirect
+	sigs.k8s.io/controller-runtime v0.1.12
+	sigs.k8s.io/controller-tools v0.1.11
+	sigs.k8s.io/testing_frameworks v0.1.1
+	sigs.k8s.io/yaml v1.1.0 // indirect
+)
+
+replace k8s.io/apimachinery => k8s.io/apimachinery v0.0.0-20190221213512-86fb29eff628
```
Comment (detiber) [inline go.mod]: Any particular reason for the apimachinery replace here?

Comment (vincepri) [reply go.mod]: Without it, go modules picks the latest version. This is a known issue with client-go pre-v1.12, seems the replace forces the older version and it works fine after that

Comment (vincepri) [reply go.mod]: For reference: https://github.com/kubernetes/client-go/blob/master/INSTALL.md#go-modules

File: Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (chuckha) [reply Dockerfile]: fine with me

**Pull request merged**